tests/ui/linking/executable-no-mangle-strip.rs

```diff
@@ -0,0 +1,22 @@
+//@ build-pass
+
+// Regression test for <https://github.com/rust-lang/rust/issues/139744>.
+// Functions in the binary marked with no_mangle should be GC-ed if they
+// are not indirectly referenced by main.
+
+unsafe extern "C" {
+    unsafe static THIS_SYMBOL_SHOULD_BE_UNDEFINED: usize;
+}
+
+#[unsafe(no_mangle)]
+pub unsafe fn used() {
+    println!("THIS_SYMBOL_SHOULD_BE_UNDEFINED = {}", unsafe { THIS_SYMBOL_SHOULD_BE_UNDEFINED });
+}
```

Comment:
Nit: I know that `#[no_mangle]` implies it, but could we also add an explicit test for `#[used]`?

Reply:
Using `#[used]` emits errors, while `#[no_mangle]` does not, on MacOS.

Don't know why.

Reply:
I know why now.

On MacOS, `#[used]` emits `llvm.compiler.used`. So `#[used(compiler)]` is needed here. Is it expected behavior?

Reply:
TIL that `#[used]` is an alias for `#[used(linker)]` on some platforms: https://github.com/rust-lang/rust/blob/f836ae4e663b6e8938096b8559e094d18361be55/compiler/rustc_codegen_ssa/src/codegen_attrs.rs#L195-L223 Eventually `#[used]` should be changed to `#[used(linker)]` unconditionally anyway.